$MirOS: ports/sysutils/ntfsprogs/patches/patch-libntfs_device_c,v 1.1.7.1 2005/03/18 15:52:18 tg Exp $
$NetBSD: patch-ae,v 1.1.1.1 2004/10/05 12:39:38 agc Exp $
--- libntfs/device.c.orig	Fri Feb 24 15:49:36 2006
+++ libntfs/device.c	Sun Apr 29 22:39:26 2007
@@ -85,6 +85,10 @@
 #	define BLKBSZSET _IOW(0x12,113,size_t) /* Set device block size in bytes. */
 #endif
 
+#if defined(__NetBSD__) || defined(__OpenBSD__)
+#include <sys/disklabel.h> /* XXX autoconf this ? */
+#endif
+
 /**
  * ntfs_device_alloc - allocate an ntfs device structure and pre-initialize it
  * @name:	name of the device (must be present)
@@ -528,6 +532,23 @@ s64 ntfs_device_size_get(struct ntfs_dev
 					(unsigned long)this_floppy.size,
 					(unsigned long)this_floppy.size);
 			return (s64)this_floppy.size * 512 / block_size;
+		}
+	}
+#endif
+#ifdef DIOCGPART
+	{
+		struct stat st;
+		if (dev->d_ops->stat(dev, &st) >= 0) {
+			struct disklabel disklabel;
+			int secsize;
+			s64 psize;
+			if (dev->d_ops->ioctl(dev, DIOCGDINFO, &disklabel) >= 0) {
+				secsize = disklabel.d_secsize;
+				psize = disklabel.d_partitions[DISKPART(st.st_rdev)].p_size;
+				ntfs_log_debug("DIOCGPART nr %d byte blocks = %lld (0x%llx)\n",
+						secsize, psize, psize);
+				return psize * secsize / block_size;
+			}
 		}
 	}
 #endif
